### ISSUE TITLE
ESP32C3 : not able to wake up HIGH level on GPIO 0 or 1

### ISSUE THREAD

**Issue by spilz87** [espressif/arduino-esp32#8447]
### Board

ESP32-C3

### Device Description

custom board with ESP32-C3

### Hardware Configuration

rotary encoder connected to GPIO 0 and 1 with 100k PULLUP resistor connected to 3.3v

### Version

latest master (checkout manually)

### IDE Name

Arduino IDE

### Operating System

Windows 10

### Flash frequency

80MHz

### PSRAM enabled

yes

### Upload speed

921600

### Description

I'm building a low power remote.
I use a rotary encoder connected to GPIO 0 and 1.
I would like to use it to wake up the ESP32C3, sending data and going back to deep sleep.
the goal is to be able to use a small battery and not having to charge it for one month (or more)

I'm able to wake up from deep sleep by LOW level, but I'm not able to wake up by HIGH level.

An other information, if I put 1k PULLUP resistor, it works well but consume too much energy in deep sleep (close to 1mA)

I guess there is something I shoud disable or set to be able to use this wake up by high level.

thanks for your help in advance.

PS. : I already checked the issue #6656 but it's not the same issue and it didn't solve mine.

### Sketch

```cpp
void setup() {
  pinMode(0, INPUT);
  pinMode(1, INPUT);
  
  delay(1000);
  gpio_deep_sleep_hold_dis();
  esp_sleep_config_gpio_isolate();
  esp_deep_sleep_enable_gpio_wakeup(1<<0, ESP_GPIO_WAKEUP_GPIO_HIGH);
  esp_deep_sleep_enable_gpio_wakeup(1<<1, ESP_GPIO_WAKEUP_GPIO_HIGH);
  gpio_set_direction(GPIO_NUM_0, GPIO_MODE_INPUT);
  gpio_set_direction(GPIO_NUM_1, GPIO_MODE_INPUT);
  delay(500);
  esp_deep_sleep_start();
}

void loop() {
  // put your main code here, to run repeatedly:

}
```

### Debug Message

```plain
no debug message, I use power consumation analyser to detect if it wakes up
```

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

**Comment by mcenderdragon** [comment]
why not use a 100k pull up instead ?

**Comment by spilz87** [comment]
> why not use a 100k pull up instead ?

I put a 100k and it doesn't work 
With a 1k it works but consumes too much

**Comment by SuGlider** [comment]
I have tested this sketch with an ESP32-C3 using just a wire to shortcut GPIO 0 or 1 to 3.3v pin.
I used Arduino Core 2.0.11 and it worked fine. I waked up whenever I touched the wire to the 3.3v.

**Comment by SuGlider** [comment]
>rotary encoder connected to GPIO 0 and 1 with 100k PULLUP resistor connected to 3.3v

ESP32-C3 wakeup works by LEVEL 3.3, not by edge.
Maybe you could try it by removing any resistor.
When it is floating or on GND, it won't wakeup.
The GPIO must go to 3.3v directly.

**Comment by spilz87** [comment]
@SuGlider thanks for your reply

with a shortcut or a low resistor it works fine.

my problem is that I use a rotary encoder, which I want to use to wake up the ESP32C3.

The goal is to reduce a lot the consuption.
As I can not know in which state the rotary encoder will stop, I have to be avare about the fact both pins can be HIGH or LOW when it enter in deep sleep mode.
if one of the pin is connected to commun (GND) the value of the resistor as an impact on the deep sleep consuption.
that's why I need 100k, not a short cut.

the weird part is the first deep sleep is inpacted, not the others

**Comment by spilz87** [comment]
> > rotary encoder connected to GPIO 0 and 1 with 100k PULLUP resistor connected to 3.3v
> 
> ESP32-C3 wakeup works by LEVEL 3.3, not by edge. Maybe you could try it by removing any resistor. When it is floating or on GND, it won't wakeup. The GPIO must go to 3.3v directly.

Do you mean using the rotary encoder in the opposite way : commun connected to 3.3V ?
I need to be able to use the rotary encoder when the esp32C3 is awake too

**Comment by lbernstone** [comment]
The internal resistors in an esp32 are about 50-60K.  I suspect you need to overcome that resistance to toggle the interrupt.  Try smaller resistors until you get it to wake.

**Comment by SuGlider** [comment]
@spilz87 - I'm sorry but I'm not following you, please send me some diagram or hand drawing of how it is all connected. Thanks.

**Comment by spilz87** [comment]
I think the issue is directly related to #8449

**Comment by SuGlider** [comment]
@spilz87 - The C3 RTC GPIO wake up needs no pullup. Please try it with no resistors.
https://github.com/espressif/arduino-esp32/issues/8449#issuecomment-1664706413

**Comment by VojtechBartoska** [comment]
any updates on this @spilz87?

**Comment by spilz87** [comment]
> any updates on this @spilz87?

I wait to be home to test again

do you have the same issue ?

**Comment by VojtechBartoska** [comment]
No, I didn't test, just asking about the status. :) We will appreciate if you can retest this, ideally on 2.0.11 version. Thanks

**Comment by spilz87** [comment]
I did the tests and give more informations in this issue #8449

**Comment by spilz87** [comment]
rotary encoder doesn't work properly without pullup nor pulldown resistor.
on 2.0.10 and 2.0.11 it become even worth.

my hypotesis : it is linked to internal pulldown resistor which seems enabled when enter in deep sleep mode.

**Comment by spilz87** [comment]
> @spilz87 - I'm sorry but I'm not following you, please send me some diagram or hand drawing of how it is all connected. Thanks.

Here is the diagram used : 
![Schema Rotary encoder](https://github.com/espressif/arduino-esp32/assets/20712939/6f2ba3ac-5018-4ace-b3c4-0be9b46d37d0)

thanks for your support

**Comment by VojtechBartoska** [comment]
thanks for more info @spilz87, Can you please take a look @SuGlider?

**Comment by spilz87** [comment]
Hi @SuGlider 
Did you have time to have a look ? 
Thanks in advance 
Have a good day

**Comment by SuGlider** [comment]
Hi, not yet, I don't have a rotary in order to test it better. I just have a few dupont cables.

**Comment by SuGlider** [comment]
> > @spilz87 - I'm sorry but I'm not following you, please send me some diagram or hand drawing of how it is all connected. Thanks.
> 
> Here is the diagram used : ![Schema Rotary encoder](https://user-images.githubusercontent.com/20712939/261672401-6f2ba3ac-5018-4ace-b3c4-0be9b46d37d0.JPG)
> 
> thanks for your support

I understand that, based  on the circuit, even if the ESP32-C3 is not present, it will consume current because of the rotary pullup resistors.
The ROTSW_A or ROTSW_B pins are always connected to GND (one or the other). This makes a permanent circuit from VCC to GND through the resistors.

**Comment by SuGlider** [comment]
Not sure why would this part of the code be necessary:

``` cpp
  gpio_deep_sleep_hold_dis();
  esp_sleep_config_gpio_isolate();
```

@spilz87 - is it an attempt to reduce the electrical current consumption?

**Comment by spilz87** [comment]
> > > @spilz87 - I'm sorry but I'm not following you, please send me some diagram or hand drawing of how it is all connected. Thanks.
> 
> > 
> 
> > Here is the diagram used : ![Schema Rotary encoder](https://user-images.githubusercontent.com/20712939/261672401-6f2ba3ac-5018-4ace-b3c4-0be9b46d37d0.JPG)
> 
> > 
> 
> > thanks for your support
> 
> 
> 
> I understand that, based  on the circuit, even if the ESP32-C3 is not present, it will consume current because of the rotary pullup resistors.
> 
> The ROTSW_A or ROTSW_B pins are always connected to GND (one or the other). This makes a permanent circuit from VCC to GND through the resistors.

Hi,
the basic rotary encoder act like switchs between A and COM, and B and COM. In my schematic COM is GROUND. 
In stable position : 
A and B are low (switchs closed)
Or 
A and B are high (switchs opened)

In mouvement : 
Depends the direction, A or B changes then the other one follows. 

So yes, in some position (switchs closed) there is a permanent consumption, that's why it is necessary to increase the resistor. 

As it seems the internal pull down resistors are activated at deep sleep mode, only small resistor connected to rotary encoder can compensate and so big consumption for nothing. 

Tell me if it's not clear :)

**Comment by SuGlider** [comment]
The problem that I see is that the Encoder KY-040 (mechanical Arduino Rotary Encoder) causes bouncing signals, like a mechanical push button.

Most probably, what you call "too fast" is the noise.

**Comment by spilz87** [comment]
Yes, in use, the encoder causes this kind of bouncing signals, that's why the external pull-up resistors are needed.

But I'm talking about waking up from deep sleep.
I just need to wake up properly, all the code is working well.

The issue seems to occure when enter in deep sleep mode, so when encoder is not moving any more for few secondes

**Comment by SuGlider** [comment]
I'm a bit lost already. What is exactly the issue you want to solve?

**Comment by spilz87** [comment]
> I'm a bit lost already. What is exactly the issue you want to solve?

When I enter in deep sleep mode and GPIO (0 or 1) are LOW level (rotary encoder "closed"), I'm not able to wake up when I turn the encoder (rotary encoder "opened").

**Comment by spilz87** [comment]
Experiment : 
- External pullup resistor 100ko
- rotary encoder in fixed position "closed"
- GPIO 1 meserue 0v -> logic level LOW
- Enter in deep sleep mode
- Turn the rotary encoder (go to ->"open" state)
- GPIO 1 meserue 0.4v/0.5v -> it seems the internal pulldown resistor is enabled and interfer with the external pullup resistor. -> logic level still LOW -> not detected -> no wake up

On version 2.0.9 it reacted like this only on the first enter in deep sleep mode, so I found a turn arround 
On version 2.0.10 and 2.0.11 it reacts always like this.

according to this different reactions depends on the version, it clearly indicates a driver issue for me

**Comment by SuGlider** [comment]
@spilz87 - I don't have a Rotary Encoder here to test it, but I have used a push button.
In this case I just used 3.3v connected to one end of the push button and the other end going to the GPIO 0.
The same for a second push button connected to GPIO 1.
Whenever necessary, I switch the wires to GND.

No external pullup, not set to input mode, no isolation mode, no hold GPIO disabled.
GPIO ---> BUTTON --> 3.3V | GND 

It works fine. Why wouldn't it work with a Rotary Encoder?
I'm using Arduino 2.0.11

The sketch used is this:
``` cpp
RTC_DATA_ATTR int bootCount = 0;
#define GPIO_A 0
#define GPIO_B 0

void print_wakeup_reason(){
  esp_sleep_wakeup_cause_t wakeup_reason;

  wakeup_reason = esp_sleep_get_wakeup_cause();

  switch(wakeup_reason)
  {
    case ESP_SLEEP_WAKEUP_EXT0 : Serial.println("Wakeup caused by external signal using RTC_IO"); break;
    case ESP_SLEEP_WAKEUP_EXT1 : Serial.println("Wakeup caused by external signal using RTC_CNTL"); break;
    case ESP_SLEEP_WAKEUP_TIMER : Serial.println("Wakeup caused by timer"); break;
    case ESP_SLEEP_WAKEUP_TOUCHPAD : Serial.println("Wakeup caused by touchpad"); break;
    case ESP_SLEEP_WAKEUP_ULP : Serial.println("Wakeup caused by ULP program"); break;
    case ESP_SLEEP_WAKEUP_GPIO : Serial.println("Wakeup caused by RISC V GPIO"); break;
    default : Serial.printf("Wakeup was not caused by deep sleep: %d\n",wakeup_reason); break;
  }
}


void setup() {
  Serial.begin(115200);
  pinMode(GPIO_A, INPUT);
  pinMode(GPIO_B, INPUT);
  Serial.println("\nSTARTING...\n");
  
  delay(1000);
  //Increment boot number and print it every reboot
  ++bootCount;
  Serial.println("Boot number: " + String(bootCount));
  Serial.println();
  //Print the wakeup reason for ESP32
  print_wakeup_reason();

  esp_deepsleep_gpio_wake_up_mode_t GPIO_A_mode = ESP_GPIO_WAKEUP_GPIO_HIGH;
  esp_deepsleep_gpio_wake_up_mode_t GPIO_B_mode = ESP_GPIO_WAKEUP_GPIO_HIGH;
  if (digitalRead(GPIO_A)) {
     GPIO_A_mode = ESP_GPIO_WAKEUP_GPIO_LOW;
  }
  if (digitalRead(GPIO_B)) {
     GPIO_B_mode = ESP_GPIO_WAKEUP_GPIO_LOW;
  }

  esp_deep_sleep_enable_gpio_wakeup(1<<GPIO_A, GPIO_A_mode);
  esp_deep_sleep_enable_gpio_wakeup(1<<GPIO_B, GPIO_B_mode);

  delay(500);
  Serial.println("Sleep Begining.");
  esp_deep_sleep_start();
}

void loop() {}
```

**Comment by SuGlider** [comment]
This is the breadboard circuit:

![image](https://github.com/espressif/arduino-esp32/assets/65977697/29af03d2-a17d-4dd1-b398-9f8a890db993)

**Comment by SuGlider** [comment]
If I keep pressed the Yellow Button, the S3 wakes up and returns to sleep, not waking up again (meaning that the state changed to HIGH making it wake up and then the sketch only wakes up again on LOW).
Then I relese the button and move the red wire to GND pin from the C3 board.
When I press it again, the C3 wakes up (on LOW).

Therefore, it reads the GPIO, sets the correct state to start sleeping and wakes up when the State of the GPIO changes from what has been read to the inverse.

**Comment by spilz87** [comment]
did you put the 100ko resistor ?

**Comment by SuGlider** [comment]
I understant that the Rotary Encoder pins A and B are always on GND or on VCC (3.3). Never floating.
Right?

**Comment by spilz87** [comment]
so you can not be impacted by internal pulldown resistor

**Comment by SuGlider** [comment]
No, but it works.... This is what matters, isn't it?

**Comment by spilz87** [comment]
> I understant that the Rotary Encoder pins A and B are always on GND or on VCC (3.3). Never floating. Right?

no, it acts like a switch with ground, so you need pullup resistor otherwile you would have a short cut

**Comment by SuGlider** [comment]
> > I understant that the Rotary Encoder pins A and B are always on GND or on VCC (3.3). Never floating. Right?
> 
> no, it acts like a switch with ground, so you need pullup resistor otherwile you would have a short cut

I don't think so.... the GPIO is in INPUT mode, no pull up/down. 
It is reading, not outputing any level.

**Comment by SuGlider** [comment]
Anyway, the C3 doesn't reset or anything. With a short cut, it must reset.

**Comment by spilz87** [comment]
> > > I understant that the Rotary Encoder pins A and B are always on GND or on VCC (3.3). Never floating. Right?
> > 
> > 
> > no, it acts like a switch with ground, so you need pullup resistor otherwile you would have a short cut
> 
> I don't think so.... the GPIO is in INPUT mode, no pull up/down. It is reading, not outputing any level.

actually I tell you

**Comment by spilz87** [comment]
> Anyway, the C3 doesn't reset or anything. With a short cut, it must reset.

because you are not in the same configuration

put a 100ko pullup resistor between your switch and 3.3v and when you press connect to the ground

**Comment by SuGlider** [comment]
I understant that you want the Rotary Encoder to wake up the C3. Is it correct?

**Comment by spilz87** [comment]
![switch](https://github.com/espressif/arduino-esp32/assets/20712939/dee50ed7-a42b-4074-bf18-20f699cbae17)

In this configuration, I want the ESP wakes up when switch pass from close to open

**Comment by SuGlider** [comment]
OK, I'll set my circuit to use the resistor. Hold on.

**Comment by spilz87** [comment]
thanks :)

**Comment by SuGlider** [comment]
Using the resistor, the sketch doesn't work because it reads 2.75v and doesn't wake up the C3.
It only wakes up on 3.3v.

The Rotary Circuit is not the same of the button:
```
VCC -----+
         |---- GPIO (it does to VCC or to GND, but VCC never connects to GND (no short cut).
 GND---- + 
 ```
 
![image](https://github.com/espressif/arduino-esp32/assets/65977697/91074f35-bd87-40f6-b6cd-f0bd6af93d8e)

Output A is in the grey or orange area, never in both. Each area is connected to GND or VCC.

**Comment by SuGlider** [comment]
I can't understand why you need a 100K resistor.... there is no short cut.

**Comment by spilz87** [comment]
![rot](https://github.com/espressif/arduino-esp32/assets/20712939/3862056e-68c9-4430-958b-a658c2ef0955)
a rotary encoder works like this

[https://www.youtube.com/watch?app=desktop&v=ns1iqNInzG8](url)

**Comment by spilz87** [comment]
> Using the resistor, the sketch doesn't work because it reads 2.75v and doesn't wake up the C3. It only wakes up on 3.3v.
> 
> The Rotary Circuit is not the same of the button:
> 
> ```
> VCC -----+
>          |---- GPIO (it goes to VCC or to GND, but VCC never connects to GND (no short cut).
>  GND---- + 
> ```
> 
> ![image](https://user-images.githubusercontent.com/65977697/263830176-91074f35-bd87-40f6-b6cd-f0bd6af93d8e.png)
> 
> Output A is in the grey or orange area, never in both. Each area is connected to GND or VCC.

in your picture, it explains : 
- A and B are the small round parts
- Common are all the big orange parts
when you turn, the Common parts turn and are in contact with A or B then with both.
When Common doesnt touch A ( or B) the circuit is opened 
When Common touchs A ( or B) the cicuit is closed between Common and A (or Common and B)

**Comment by SuGlider** [comment]
Yes, in the Orange/Grey picture, there is no pull up. It is directly connected to VCC or GND.
Output A will be directly connected to VCC or to GND.

If the device works in this direct way, the C3 will wake up.
If the deveice has internally a resistor, it won't.

**Comment by spilz87** [comment]
> Yes, in the Orange/Grey picture, there is no pull up. It is directly connected to VCC or GND. Output A will be directly connected to VCC or to GND.
> 
> If the device works in this direct way, the C3 will wake up. If the deveice has internally a resistor, it won't.

I don't get why you say it is directly connected to VCC or GND
on this picture it just show you the mecanical part, not a schemat

**Comment by SuGlider** [comment]
Can you use a multimeter and be able to measure such internal resistor in the Rotary Encoder?

**Comment by spilz87** [comment]
> Can you use a multimeter and be able to measure such internal resistor in the Rotary Encoder?

it's like a switch : 0 or Infini

**Comment by SuGlider** [comment]
So there is no resistance?

**Comment by spilz87** [comment]
> So there is no resistance?

no, like a switch

**Comment by SuGlider** [comment]
If you use the multimeter with one end to the OutputA and the other end to the VCC, it will read Zero ohms.
Therefore there is no internal "hidden" resistor.

In that way, the Output A is exlusivelly directly connected to the VCC or GND pins of the Rotary Encoder, right?

**Comment by SuGlider** [comment]
If this is correct, conneting VCC and GND to the respective Rotary Encoder Pins, and OutputA to GPIO 0, thre is no short cut.
Is that right, isn't it?

**Comment by spilz87** [comment]
> If you use the multimeter with one end to the OutputA and the other end to the VCC, it will read Zero ohms. Therefore there is no internal "hidden" resistor.
> 
> In that way, the Output A is exlusivelly directly connected to the VCC or GND pins of the Rotary Encoder, right?

No, you mesure resistor without power connected

**Comment by spilz87** [comment]
> If this is correct, conneting VCC and GND to the respective Rotary Encoder Pins, and OutputA to GPIO 0, thre is no short cut. Is that right, isn't it?

you will get shortcut

**Comment by SuGlider** [comment]
> > If you use the multimeter with one end to the OutputA and the other end to the VCC, it will read Zero ohms. Therefore there is no internal "hidden" resistor.
> > In that way, the Output A is exlusivelly directly connected to the VCC or GND pins of the Rotary Encoder, right?
> 
> No, you mesure resistor without power connected

What is the resistance measeured?

**Comment by spilz87** [comment]
the resistance (ohm) is a passif measure 
you can not mesure the resistance when it's connected
you can mesure only tension (Volt) or current (Amps) when your system is working

**Comment by spilz87** [comment]
![rott](https://github.com/espressif/arduino-esp32/assets/20712939/99165d81-97ea-4c00-9e74-a13b7c966b7c)
as you can see, the common representation of a rotary encoder is 2 switchs and we always add 2 pull up resistors

**Comment by SuGlider** [comment]
You must be able to measure the resistance using just the encoder with nothing connected to it.
Connect the Multimeter to the Encoder VCC pin and A pin with no power on it.
It must read a resistor R.

**Comment by spilz87** [comment]
> You must be able to measure the resistance using just the encoder with nothing connected to it. Connect the Multimeter to the Encoder VCC pin and A pin with no power on it. It must read a resistor R.

when you say VCC you mean Common ?

**Comment by SuGlider** [comment]
Please send me a picture of your encoder.

**Comment by SuGlider** [comment]
Is it like this picture?

![image](https://github.com/espressif/arduino-esp32/assets/65977697/b123f0e3-63dd-4ea8-b8ca-084b0b00a387)

**Comment by spilz87** [comment]
> Is it like this picture?
> 
> ![image](https://user-images.githubusercontent.com/65977697/263837228-b123f0e3-63dd-4ea8-b8ca-084b0b00a387.png)

No
this one as already pull up resistor on the board

**Comment by SuGlider** [comment]
Could you find a picture of your encoder in the Web and post here?

**Comment by spilz87** [comment]
![rot3](https://github.com/espressif/arduino-esp32/assets/20712939/edce1445-5c46-4c9c-b4a1-369aa99f9830)

**Comment by spilz87** [comment]
![rot4](https://github.com/espressif/arduino-esp32/assets/20712939/4d6b0c3a-5d43-4263-b49b-f5c195695aed)

explanation from [https://how2electronics.com/construction-working-rotary-encoder/](url)

**Comment by SuGlider** [comment]
Is your Rotary Encoder like this one (with eletronic components - resistors and capacitors)

![image](https://github.com/espressif/arduino-esp32/assets/65977697/35d82707-20ec-4992-ab65-49f8e9823215)

**Comment by spilz87** [comment]
> Is your Rotary Encoder like this one (with eletronic components - resistors and capacitors)
> 
> ![image](https://user-images.githubusercontent.com/65977697/263838778-35d82707-20ec-4992-ab65-49f8e9823215.png)

no, it is mounted in a custom board with pullup resistor

**Comment by SuGlider** [comment]
Ok, it won't work.
The C3 need a full 3.3v in order to wake up.

**Comment by spilz87** [comment]
> Ok, it won't work. The C3 need a full 3.3v in order to wake up.

Yes I agree
if there is no pulldown resistor you got a 3.3V with the pullup resistor

**Comment by SuGlider** [comment]
I think that the "simpler/cheap encoder model" like in the [picture](https://github.com/espressif/arduino-esp32/issues/8447#issuecomment-1696430011) from would worj fine with no Pullup/down resistor.

**Comment by spilz87** [comment]
> Ok, it won't work. The C3 needs a full 3.3v in order to wake up. The pull up 100K resistors will reduce the voltage, making a voltage divisor.

Normaly no.
if there is no pull down resistor, as the input consumes almost 0 amps, there is no current through the resistor so according to U = RI, the Tention between both side of the resistance is 0V so it acts like GPIO is conected to 3.3V

**Comment by spilz87** [comment]
> I think that the "simpler/cheap encoder model" like in the [picture](https://github.com/espressif/arduino-esp32/issues/8447#issuecomment-1696430011) from would work fine with no Pullup/down resistor.

this is not the issue.

**Comment by SuGlider** [comment]
Ok, so what is the issue?

**Comment by spilz87** [comment]
> ![switch](https://user-images.githubusercontent.com/20712939/263823329-581860df-0480-4d7d-87fe-8a0c7f98f243.JPG)
> 
> In this configuration, I want the ESP wakes up when switch open

this is the simplest way to explain the issue I think

**Comment by spilz87** [comment]
when I open the switch, the ESP should wake up, but it doesn't

**Comment by SuGlider** [comment]
It doesn't work as you aready know. I measured the voltage in GPIO0 and it is 2.75v or 0v.
2.75v doesn't wake up the C3.

**Comment by spilz87** [comment]
> It doesn't work as you aready know. I measured the voltage in GPIO0 and it is 2.75v or 0v. 2.75v doesn't wake up the C3.

how did you do your measure ?
with resistor ?
on deep sleep ?
when ESP is awke ?

**Comment by SuGlider** [comment]
With a 100K resistor in R1 (instead of the 10K from the picture).
I measured it with the ESP32 waked up.

**Comment by spilz87** [comment]
when ESP32 wake up, I measure 3.3V

did you put the pin in input mode ?

**Comment by spilz87** [comment]
if you connect the GPIO to 3.3V with a resistor and nothing else (no pull down resistor) you should read 3.3V.
this is one of the main caracteritics of microcontrolers. That's why we can read the logic position of a switch HIGH or LOW.

**Comment by spilz87** [comment]
> With a 100K resistor in R1 (~instead of the 10K from the picture~). Exactly like in the picture. I measured it with the ESP32 waked up.

can I get a picture ?

**Comment by SuGlider** [comment]
I just tested it with 100R, 1K, 10K and 100K pullup.
All of them work, but the 100K.

So, there is a problem with 100K pullup resistance.

**Comment by spilz87** [comment]
If you try with a 1M it will not work

**Comment by spilz87** [comment]
It seems like the internal pull down resistor is enable, so it create a resistor divider. 
If I remember well, the internal pull down resistor is around 10k
So with 10k pull up resistor, you will get around 1.7v which can wake up the device

**Comment by SuGlider** [comment]
> > With a 100K resistor in R1 (~instead of the 10K from the picture~). Exactly like in the picture. I measured it with the ESP32 waked up.
> 
> can I get a picture ?

I was measuring it with the GPIO 0 in ADC mode... but it seemd that it was actually reading 3.3V (4095 in 0..4095), but the Arduino function says it was 2.75v. Therefore, it was giving me a bad number.

But it seems that with the same [circuit](https://github.com/espressif/arduino-esp32/issues/8447#issuecomment-1696448686) as in your diagram, but using a 10K resistor. 
The [sketch](https://github.com/espressif/arduino-esp32/issues/8447#issuecomment-1696300913) works as expected.

When it starts, it enters in sleep mode. 
When I pressed and hold the button down, it wakes up and then sleeps.
When I release the button, it wakes up and then sleeps. ---> go to the line before and so on.

**Comment by spilz87** [comment]
And to remember
With V2.0.9 we don't have the same result as we have with V2.0.10 and V2.0.11

**Comment by SuGlider** [comment]
> With V2.0.9 we don't have the same result as we have with V2.0.10 and V2.0.11.

How is it different?

**Comment by spilz87** [comment]
> > With V2.0.9 we don't have the same result as we have with V2.0.10 and V2.0.11.
> 
> 
> 
> How is it different?

With V2.0.9 it happens only at the first enter in deep sleep mode, but then it disappear. 
So if when you enter in the first sleeping mode you put a wake up interrupt on LOW level, it will wake up immediately and then you can put the wake up interrupt on HIGH level and it works fine. 

On V2.0.10 and V2.0.11 it never works when you try to wake up from HIGH level. 

Tests did with the same board, same sketch, just changing the version from V2.0.9 to V2.0.10 and V2.0.11

This difference for me is an argue to say it linked to software/library.

**Comment by spilz87** [comment]
> I just tested it with 100R, 1K, 10K and 100K pullup.
> 
> All of them work, but the 100K.
> 
> 
> 
> So, there is a problem with 100K pullup resistance.
> 
> 

Did you measure the voltage each time ? (Between GND and GPIO)

**Comment by spilz87** [comment]
> > > With a 100K resistor in R1 (~instead of the 10K from the picture~). Exactly like in the picture. I measured it with the ESP32 waked up.
> 
> > 
> 
> > can I get a picture ?
> 
> 
> 
> I was measuring it with the GPIO 0 in ADC mode... but it seemd that it was actually reading 3.3V (4095 in 0..4095), but the Arduino function says it was 2.75v. Therefore, it was giving me a bad number.
> 
> 
> 
> But it seems that with the same [circuit](https://github.com/espressif/arduino-esp32/issues/8447#issuecomment-1696448686) as in your diagram, but using a 10K resistor. 
> 
> The [sketch](https://github.com/espressif/arduino-esp32/issues/8447#issuecomment-1696300913) works as expected.
> 
> 
> 
> When it starts, it enters in sleep mode. 
> 
> When I pressed and hold the button down, it wakes up and then sleeps.
> 
> When I release the button, it wakes up and then sleeps. ---> go to the line before and so on.

Do the same test with : 
- 100k resistor 
- on V2.0.9 
- You press the button before powering the board 
- You release after enter the first time in sleep mode 

It should works as for me

**Comment by spilz87** [comment]
With the sketch provided in #8445

**Comment by SuGlider** [comment]
Yes, confirmed.
Arduino Core 2.0.9 works with the sketch from #8449, even with no Pullup Resistor.
Arduino Core 2.0.11 doesn't work.

I'll find out if it has something to do Arduino Digital GPIO API or with IDF Deep Sleep API.

**Comment by SuGlider** [comment]
@VojtechBartoska / @me-no-dev 
It seems to be due to a change from IDF4.4.4 to IDF 4.4.5.

**Comment by spilz87** [comment]
@SuGlider Thanks a lot for your work 

I think there was a small issue too in IDF SDK Ver v4.4.4 as at the first time it doesn't work properly too.

**Comment by SuGlider** [comment]
@spilz87 - Yes, I saw that too. If the button is kept pressed when the ESP32-C3 + IDF 4.4.4 is booted, it may fail.
With IDF 4.4.5, it works fine, if the pullup resistor is 10K. 
This is a limitation. In order to make the Rotary Encoder wake up the C3, it shall use IDF 4.4.5 + 10k or so resistor.

You may also try IDF 5.1 that will be used with Arduino Core 3.0.0.
But I guess that it may actually be a hardware limitation that has a software workaround.

This may be a trade off... not fixable.

**Comment by spilz87** [comment]
And about using 10k instead of 100k : 
With rotary encoder I don't control the position when it enter in sleep mode, and as you saw, in some position it acts as closed switch so it drains to much curent, so it's not an available work around :(

**Comment by SuGlider** [comment]
It may require another hardware/rotary or some other circuit.
Not the best solution... maybe an addition Schmitt Trigger CI?

**Comment by SuGlider** [comment]
>So for me it's a bug not an hardware limitation.

Anyway, if you think that it is a SW bug, it shall be posted in the IDF issue instead...
Here we can't fix it. All the function calls are IDF and not Arduino related.

I think that it is a HW bug + a SW workaround in IDF 4.4.5.
There is a commit in IDF code that mentions some problem with current leaking.

**Comment by spilz87** [comment]
The board is already designed 

And it's the better way to connect it as every datasheets show
And I use it on other chip with no problem. 

This is clearly a code/driver/IDF issue that can be solve

**Comment by SuGlider** [comment]
>This is clearly a code/driver/IDF issue that can be solve

Here we just use IDF. If IDF has a bug, it shall be reported in the IDF issues.
https://github.com/espressif/esp-idf/issues

**Comment by spilz87** [comment]
> >So for me it's a bug not an hardware limitation.
> 
> 
> 
> Anyway, if you think that it is a SW bug, it shall be posted in the IDF issue instead...
> 
> Here we can't fix it. All the function calls are IDF and not Arduino related.
> 
> 
> 
> I think that it is a HW bug + a SW workaround in IDF 4.4.5.
> 
> There is a commit in IDF code that mentions some problem with current leaking.

Can you give me a link where I can post this issue for IDF ? 

Yes this issue has a direct impact on current consumption, and solving it will help a lot of people who use deep sleep mode to save battery

**Comment by SuGlider** [comment]
>Can you give me some detail which could help them to understand what you did ?

Explain the IDF version used, expected behaviour, the problem, the circuit and post, if possible, a pure IDF minimum code.
Explain how to replicate the issue with clear steps and how to see the problem (turn on, keeping the button pressed, release it etc).

>Which function/code of the IDF is called ?

Based on the issue #8449 code:
IDF calls are:
`esp_deep_sleep_enable_gpio_wakeup()` - [Documentation](https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32c3/api-reference/system/sleep_modes.html#_CPPv433esp_deep_sleep_enable_gpio_wakeup8uint64_t33esp_deepsleep_gpio_wake_up_mode_t)
`gpio_set_direction()` - [Documentation](https://docs.espressif.com/projects/esp-idf/en/release-v4.4/esp32c3/api-reference/peripherals/gpio.html#_CPPv418gpio_set_direction10gpio_num_t11gpio_mode_t)
`esp_deep_sleep_start()` - [Documentation](https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32c3/api-reference/system/sleep_modes.html#_CPPv420esp_deep_sleep_startv)

Arduino calls:
`pinMode()`
`digitalRead()`
`delay()`
`setup()`
`loop()`

IDF example for DeepSleep and GPIO wakeup:
https://github.com/espressif/esp-idf/blob/82cceabc6e6a0a2d8c40e2249bc59917cc5e577a/examples/system/deep_sleep/main/gpio_wakeup.c
https://github.com/espressif/esp-idf/blob/82cceabc6e6a0a2d8c40e2249bc59917cc5e577a/examples/system/deep_sleep/

**Comment by spilz87** [comment]
Is there a way to select which IDF is used when I compile ? 

For now I have to uninstall V2.0.11 to compile this project and reinstall to work on other project :(

**Comment by TD-er** [comment]
When using PlatformIO, you can specify your platform packages per environment you build.

**Comment by spilz87** [comment]
Hello,

I discover this on **esp_sleep.h** : 

```
 * @note You don't need to care to pull-up or pull-down before using this
 *       function, because this will be set internally in esp_deep_sleep_start
 *       based on the wakeup mode. BTW, when you use low level to wake up the
 *       chip, we strongly recommend you to add external resistors (pull-up).
```

which it's not a good idea when we want to reduce consumption (goal of deep sleep)

do we have an acces to **esp_deep_sleep_start** through arduino to modify it ?

**Comment by spilz87** [comment]
I open an issue here : 
[https://github.com/espressif/esp-idf/issues/12183](url)

**Comment by P-R-O-C-H-Y** [comment]
Quick update:
Fix have been merged to 5.1 branch that will be used for ESP Arduino Core 3.0.0.

**Comment by VojtechBartoska** [comment]
Hello, closing this as solved. If needed, you can reopen.
